test/local/devices_tests.js

```diff
@@ -0,0 +1,196 @@
+            t.deepEqual(args[1], sessionToken.tokenId, 'second argument was sessionTokenId')
+            t.equal(args[2], device, 'third argument was device')
+
+            t.equal(log.activityEvent.callCount, 1, 'log.activityEvent was called once')
+            args = log.activityEvent.args[0]
+            t.equal(args.length, 3, 'log.activityEvent was passed three arguments')
+            t.equal(args[0], 'device.created', 'first argument was event name')
+            t.equal(args[1], request, 'second argument was request object')
+            t.deepEqual(args[2], {
+              uid: sessionToken.uid.toString('hex'),
+              device_id: deviceId.toString('hex')
+            }, 'third argument contained uid and device_id')
+
+            t.equal(log.event.callCount, 1, 'log.event was called once')
+            args = log.event.args[0]
+            t.equal(args.length, 3, 'log.event was passed three arguments')
+            t.equal(args[0], 'device:create', 'first argument was event name')
+            t.equal(args[1], request, 'second argument was request object')
+            t.deepEqual(args[2], {
+              uid: sessionToken.uid,
+              id: deviceId,
+              type: device.type,
+              timestamp: deviceCreatedAt
+            }, 'third argument was event data')
+
+            t.equal(push.notifyDeviceConnected.callCount, 1, 'push.notifyDeviceConnected was called once')
+            args = push.notifyDeviceConnected.args[0]
+            t.equal(args.length, 3, 'push.notifyDeviceConnected was passed three arguments')
+            t.equal(args[0], sessionToken.uid, 'first argument was uid')
+            t.equal(args[1], device.name, 'second arguent was device name')
+            t.equal(args[2], deviceId.toString('hex'), 'third argument was device id')
+          })
+          .then(function () {
+            db.createDevice.reset()
+            push.notifyDeviceConnected.reset()
+            log.activityEvent.reset()
+            log.event.reset()
+          })
+      }, t)
+
+      test('update', function (t) {
+        var deviceInfo = {
+          id: deviceId,
+          name: device.name,
+          type: device.type
+        }
+        return devices.upsert(request, sessionToken, deviceInfo)
+          .then(function (result) {
+            t.equal(result, deviceInfo, 'result was correct')
+
+            t.equal(db.createDevice.callCount, 0, 'db.createDevice was not called')
+
+            t.equal(db.updateDevice.callCount, 1, 'db.updateDevice was called once')
+            var args = db.updateDevice.args[0]
+            t.equal(args.length, 3, 'db.createDevice was passed three arguments')
+            t.deepEqual(args[0], sessionToken.uid, 'first argument was uid')
+            t.deepEqual(args[1], sessionToken.tokenId, 'second argument was sessionTokenId')
+            t.equal(args[2], deviceInfo, 'third argument was device info')
```

Comment:
What Sean said. I'm asserting that deviceInfo is passed through directly in this case. I could make an extra assertion using `deepEqual` but it would be tautological to make that against `deviceInfo` because they're literally the same object. Instead I'd make it against an object literal to ensure that `deviceInfo` is unmodified, which I'm happy to do. I'll add it presently.